pkg/controller/nodeipam/ipam/multi_cidr_range_allocator.go

```diff
@@ -0,0 +1,1158 @@
+	recorder              record.EventRecorder
+	// queue is where incoming work is placed to de-dup and to allow "easy"
+	// rate limited requeues on errors
+	queue workqueue.RateLimitingInterface
+
+	// lock guards nodesInProcessing and cidrMap to avoid races in CIDR allocation.
+	lock *sync.Mutex
+	// nodesInProcessing is a set of nodes that are currently being processed.
+	nodesInProcessing sets.String
+	// cidrMap maps ClusterCIDR labels to internal ClusterCIDR objects.
+	cidrMap map[string][]*cidrset.ClusterCIDR
+}
+
+// NewMultiCIDRRangeAllocator returns a CIDRAllocator to allocate CIDRs for node (one for each ip family).
+// Caller must always pass in a list of existing nodes to the new allocator.
+// NodeList is only nil in testing.
+func NewMultiCIDRRangeAllocator(
+	client clientset.Interface,
+	nodeInformer informers.NodeInformer,
+	clusterCIDRInformer networkinginformers.ClusterCIDRInformer,
+	allocatorParams CIDRAllocatorParams,
+	nodeList *v1.NodeList,
+	testCIDRMap map[string][]*cidrset.ClusterCIDR,
+) (CIDRAllocator, error) {
+	if client == nil {
+		klog.Fatalf("client is nil")
+	}
+
+	eventBroadcaster := record.NewBroadcaster()
+	eventSource := v1.EventSource{
+		Component: "multiCIDRRangeAllocator",
+	}
+	recorder := eventBroadcaster.NewRecorder(scheme.Scheme, eventSource)
+	eventBroadcaster.StartStructuredLogging(0)
+	klog.V(0).Infof("Started sending events to API Server. (EventSource = %v)", eventSource)
+
+	eventBroadcaster.StartRecordingToSink(
+		&v1core.EventSinkImpl{
+			Interface: client.CoreV1().Events(""),
+		})
+
+	if client.CoreV1().RESTClient().GetRateLimiter() != nil {
+		ratelimiter.RegisterMetricAndTrackRateLimiterUsage("multi_cidr_range_allocator", client.CoreV1().RESTClient().GetRateLimiter())
+	}
+
+	ra := &multiCIDRRangeAllocator{
+		client:                client,
+		nodeLister:            nodeInformer.Lister(),
+		nodesSynced:           nodeInformer.Informer().HasSynced,
+		clusterCIDRLister:     clusterCIDRInformer.Lister(),
+		clusterCIDRSynced:     clusterCIDRInformer.Informer().HasSynced,
+		nodeCIDRUpdateChannel: make(chan multiCIDRNodeReservedCIDRs, cidrUpdateQueueSize),
+		recorder:              recorder,
+		queue:                 workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "multi_cidr_range_allocator"),
+		lock:                  &sync.Mutex{},
+		nodesInProcessing:     sets.NewString(),
+		cidrMap:               make(map[string][]*cidrset.ClusterCIDR, 0),
+	}
+
+	// testCIDRMap is only set for testing purposes.
+	if len(testCIDRMap) > 0 {
+		ra.cidrMap = testCIDRMap
+		klog.Warningf("testCIDRMap should only be set for testing purposes, if this is seen in production logs, it might be a misconfiguration or a bug.")
+	}
+
+	ccList, err := listClusterCIDRs(client)
+	if err != nil {
+		return nil, err
+	}
+
+	if ccList == nil {
+		ccList = &networkingv1alpha1.ClusterCIDRList{}
+	}
+	createDefaultClusterCIDR(ccList, allocatorParams)
```

Comment:
it is better if you don't mutate the list, and use the function to generate the DefaultClusterCIDR if is not in the list

```
if DefaultClusterCIDR not in ccList {
    defaultClusterCIDR := createDefaultClusterCIDR(allocatorParams)
    // I will also consider to create the defaultClusterCIDR object here to not special 
   // case // createClusterCIDR createClusterCIDR(clusterCIDR *networkingv1alpha1.ClusterCIDR, terminating bool)
    ccList = append(ccList, defaultClusterCIDR)
}
```
https://github.com/kubernetes/kubernetes/pull/109090/commits/2379e60aa66c43bf193680d650a15858570d6608#r938845559